WelsonJS.Toolkit/WelsonJS.Launcher/Program.cs

```diff
@@ -56,14 +70,165 @@ static void Main()
+                    workingDirectory: workingDirectory,
+                    entryFileName: "app.js",
+                    scriptName: "bootstrap",
+                    isConsoleApplication: true,
+                    isInteractiveServiceApplication: false
+                );
+                return;
+            }
+
+            throw new NotSupportedException($"Unsupported file type: {fileExtension}");
+        }
+
+        private static string GetAppRootDirectory()
+        {
+            string[] candidates = new[]
+            {
+                GetAppConfig("AppRootDirectory"),
+                AppDomain.CurrentDomain.BaseDirectory,
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86), "WelsonJS"),
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "WelsonJS"),
+            };
+
+            foreach (string dir in candidates)
+            {
+                if (string.IsNullOrEmpty(dir))
+                    continue;
+
+                string appJs = Path.Combine(dir, "app.js");
+                if (File.Exists(appJs))
+                    return dir;
+            }
+
+            throw new FileNotFoundException("Could not locate app.js in any known application root directory.");
+        }
+
+        private static string CreateInstanceDirectory()
+        {
+            string instanceId = Guid.NewGuid().ToString();
+            string workingDirectory = GetWorkingDirectory(instanceId);
+
+            try
+            {
+                // check if the working directory exists
+                if (Directory.Exists(workingDirectory))
+                {
+                    throw new InvalidOperationException("GUID validation failed. Directory already exists.");
+                }
+
+                Directory.CreateDirectory(workingDirectory);
+            }
+            catch
+            {
+                throw new Exception("Instance Initialization failed");
+            }
+
+            return workingDirectory;
+        }
+
+        private static void CopyDirectoryRecursive(string sourceDir, string destDir)
+        {
+            if (!Directory.Exists(sourceDir))
+            {
+                throw new DirectoryNotFoundException("Source directory not found: " + sourceDir);
+            }
+
+            Directory.CreateDirectory(destDir);
+
+            foreach (var file in Directory.GetFiles(sourceDir, "*", SearchOption.AllDirectories))
+            {
+                string relativePath = file.Substring(sourceDir.Length).TrimStart(
+                    Path.DirectorySeparatorChar,
+                    Path.AltDirectorySeparatorChar
+                );
+
+                string targetPath = Path.Combine(destDir, relativePath);
+                string targetDir = Path.GetDirectoryName(targetPath);
+                if (!Directory.Exists(targetDir))
+                {
+                    Directory.CreateDirectory(targetDir);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Path traversal vulnerability detected. User-controlled input is used in a file system operation without proper validation, which could allow an attacker to access files outside the intended directory. Validate and sanitize user input before using it in file paths, and use `Path.GetFullPath()` with `StartsWith()` checks to ensure the resulting path is within the intended directory.https://cwe.mitre.org/data/definitions/22.html